File: app/src/main/java/mydebts/android/app/feature/events/EventsActivity.java
package mydebts.android.app.feature.events;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.List;

import javax.inject.Inject;

import io.reactivex.disposables.Disposable;
import mydebts.android.app.R;
import mydebts.android.app.db.Event;
import mydebts.android.app.di.SubcomponentBuilderResolver;
import mydebts.android.app.feature.addevent.EventActivity;

public class EventsActivity extends AppCompatActivity {

    @Inject EventsViewModel viewModel;
    @Inject RecyclerView eventsRecyclerView;
    @Inject View emptyView;

    private Disposable eventsDisposable;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_events);

        ((EventsSubcomponent.Builder)SubcomponentBuilderResolver.resolve(this))
                .module(new EventsModule(this))
                .build().inject(this);

        eventsDisposable = viewModel.eventsSubject()
                .subscribe(this::accept);
    }

    @Override
    protected void onDestroy() {
        eventsDisposable.dispose();
        super.onDestroy();
    }

    void onAddEventClick() {
        startActivity(new Intent(this, EventActivity.class));
    }

    private void accept(List<Event> events) {
        boolean isEmpty = events == null || events.size() == 0;

        emptyView.setVisibility(isEmpty ? View.VISIBLE : View.GONE);
        eventsRecyclerView.setVisibility(isEmpty ? View.GONE : View.VISIBLE);
    }
}

File: app/src/main/java/mydebts/android/app/ui/events/EventsActivity.kt
package mydebts.android.app.ui.events

import android.arch.lifecycle.Observer
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v7.widget.RecyclerView
import android.view.Menu
import android.view.MenuItem
import android.view.View
import dagger.android.AndroidInjection
import mydebts.android.app.R
import mydebts.android.app.feature.event.EventActivity
import mydebts.android.app.feature.events.EventsViewModel
import mydebts.android.app.feature.people.PeopleActivity
import java.util.ArrayList
import javax.inject.Inject

class EventsActivity : AppCompatActivity() {

    @Inject lateinit var viewModel: EventsViewModel

    private val adapter = EventsAdapter()
    init {
        adapter.setOnItemClickListener { viewModel.onEventClick(it) }
    }

    private lateinit var eventsRecyclerView: RecyclerView
    private lateinit var emptyView: View
    private lateinit var addEventButton: FloatingActionButton

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_events)
        setTitle(R.string.title_events)
        bindViews()
        AndroidInjection.inject(this)
        bindViewModel()
    }

    override fun onCreateOptionsMenu(menu: Menu?): Boolean {
        menuInflater.inflate(R.menu.menu_events, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean =
            when (item?.itemId) {
                R.id.action_switch_to_people -> {
                    startActivity(PeopleActivity.newIntent(this))
                    true
                }
                else -> false
            }

    private fun bindViews() {
        eventsRecyclerView = findViewById(R.id.list_events)
        eventsRecyclerView.adapter = adapter

        emptyView = findViewById(R.id.text_no_events)

        addEventButton = findViewById(R.id.button_add_event)
        addEventButton.setOnClickListener { viewModel.onAddEventClick() }
    }

    private fun bindViewModel() {
        viewModel.events.observe(this, Observer {
            adapter.items = it ?: ArrayList()
            handleItemsVisibility()
        })
        viewModel.addEventNavigation.observe(this, Observer {
            it?.let { startActivity(EventActivity.newIntent(this)) }
        })
        viewModel.eventNavigation.observe(this, Observer {
            it?.let { startActivity(EventActivity.newIntent(this, it)) }
        })
    }

    private fun handleItemsVisibility() {
        val isEmpty = adapter.itemCount == 0
        emptyView.visibility = if (isEmpty) View.VISIBLE else View.GONE
        eventsRecyclerView.visibility = if (!isEmpty) View.VISIBLE else View.GONE
    }
}
